convert_grok.py

```diff
@@ -0,0 +1,458 @@
+            f"{name}.weight",
+            list(meta_tensor.shape),
+            dtype,
+            quantized_meta_tensor.nbytes,
+            tensor_ggml_type,
+        )
+        weights[name] = weight, scales
+    print("Loaded", len(weight_names), "files")
+
+    f.write_header_to_file()
+    f.write_kv_data_to_file()
+    f.write_ti_data_to_file()
+
+    # Now write actual tensor data.
+    tensor_info = []
+
+    for name in weight_names:
+        weight, scales = weights.pop(name)
+        tensor = convert_weight(name, weight, scales, config.experts)
+        tensor = maybe_permute_tensor(name, tensor, config)
+        _, tensor_ggml_type = get_dtype_and_ggml_type(tensor, ggml_type)
+        array = maybe_quantize_tensor(tensor, tensor_ggml_type).numpy()
+
+        print(
+            f"dumping {name}:",
+            f"{tensor_ggml_type.name}/{array.dtype}, {list(tensor.shape)}, {array.nbytes} bytes",
+        )
+        f.write_tensor_data(array)
+
+        tensor_info.append((name, list(tensor.shape), tensor_ggml_type.name))
+
+    try:
+        print(tabulate(tensor_info, headers=["name", "shape", "dtype"], tablefmt="psql"))
+    except NameError:
+        pass
+
+    if len(tensor_info) != len(weight_names):
+        print("Warning: not all tensors are converted")
+
+
+def from_numpy(array):
+    """Like torch.from_numpy, but handle ml_dtypes.bfloat16 too."""
+
+    if array.dtype == ml_dtypes.bfloat16:
+        return torch.from_numpy(array.view(np.uint8)).view(torch.bfloat16)
+    return torch.from_numpy(array)
+
+
+def convert_weight(name, weight, scales, experts, dtype=torch.float32, device=None):
+    # copied from https://gist.github.com/chu-tianxiang/ec310e15d56949fd0f351cb5f65ee7a1
+    weight = from_numpy(weight).to(device=device, dtype=dtype)
+    if scales is not None:
+        scale = from_numpy(scales).to(device=device, dtype=dtype)
+        # row parallel layers have sharded scale
+        if len(scale.shape) >= 2 and scale.shape[-2] != 1:
+            scale = scale[..., None, :]
+            weight = weight.view(*weight.shape[:-2], 8, -1, weight.shape[-1])
+            weight = (weight * scale).view(*weight.shape[:-3], -1, weight.shape[-1])
+        else:
+            weight = weight * scale
+
+    # Transpose linear matrix
+    if len(weight.shape) >= 2 and "token_embd" not in name:
+        weight = weight.transpose(-1, -2)
+
+    if name.endswith("ffn_gate_inp") or name.endswith("_exps"):
+        weight = weight[experts]  # gather.
+
+    return weight
+
+
+def maybe_permute_tensor(name, tensor, config):
+    def permute(weights, n_head):
+        return (
+            weights.reshape(n_head, 2, weights.shape[0] // n_head // 2, *weights.shape[1:])
+            .swapaxes(1, 2)
+            .reshape(weights.shape)
+        )
+
+    if name.endswith("attn_k"):
+        return permute(tensor, config.num_key_value_heads)
+    elif name.endswith("attn_q"):
+        return permute(tensor, config.num_attention_heads)
```

Comment:
Grok is using `LLAMA_ROPE_TYPE_NEOX`, so do not `permute`.

Reply:
Removed the call to this function and the function itself.